File: Leetcode/079. Word Search/WordSearch.java
// Leetcode 79. Word Search
// https://leetcode.com/problems/word-search/description/
// Runtime: 41ms

// 用简单的四方向搜索+辅助的访问判定表的方法做的
// 感觉又像是图相关问题
class Solution {
    boolean srch(char[][] board, char[] word, int row, int col, int index, boolean[][] visited) {
        boolean res = false;
        int maxrow = board.length - 1;
        int maxcol = board[0].length - 1;

        if (index == word.length) 
            return true;

        if (col > maxcol) 
            return false;

        if (row < 0) 
            return false;

        if (row > maxrow) 
            return false;

        if (col < 0) 
            return false;

        if (visited[row][col])
            return false;

        visited[row][col] = true;
        if (board[row][col] == word[index]) {
            if  (   srch(board, word, row, col - 1, index + 1, visited) ||
                    srch(board, word, row, col + 1, index + 1, visited) ||
                    srch(board, word, row - 1, col, index + 1, visited) ||
                    srch(board, word, row + 1, col, index + 1, visited) ) {
                return true;
            } else {
                visited[row][col] = false;
            }
        } else {
            visited[row][col] = false;
            return false;
        }
        return false;
    }

    public boolean exist(char[][] board, String word) {
        char[] w = word.toCharArray();

        int maxrow = board.length - 1;
        int maxcol = board[0].length - 1;

        boolean[][] visited = new boolean[maxrow + 1][maxcol + 1];
        for (int i = 0; i <= maxrow; i++) {
            for (int j = 0; j <= maxcol; j++) {
                visited[i][j] = false;
            }
        }

        int r = 0;
        int c = 0;
        int index = 0;

        while (r <= maxrow && c <= maxcol) {
            if (board[r][c] != w[index]) {
                c++;
                // move to next row
                if (c > maxcol) {
                    c = 0;
                    r++;
                }
            } else {
                visited[r][c] = true;
                if (    srch(board, w, r, c + 1, index + 1, visited) ||
                        srch(board, w, r, c - 1, index + 1, visited) ||
                        srch(board, w, r - 1, c, index + 1, visited) ||
                        srch(board, w, r + 1, c, index + 1, visited) ) {
                    return true;
                }
                
                // 从当前位置未能搜索到结果。重置visited进行下一轮搜索
                for (int i = 0; i <= maxrow; i++) {
                    for (int j = 0; j <= maxcol; j++) {
                        visited[i][j] = false;
                    }
                }
                c++;
                if (c > maxcol) {
                    c = 0;
                    r++;
                }
            }
        }
        return false;
    }
}

File: Leetcode/079. Word Search/word_search.cpp
// Leetcode 79. Word Search
// https://leetcode.com/problems/word-search/
// time complexity: TBD
// space complexity: O(size of board) --> for visited table

class Solution 
{
public:
    bool run(vector<vector<bool>>& visited, int idx, string& word, vector<vector<char>>& board, int r, int c)
    {
        int row = board.size();
        int col = board[0].size();
        
        if (r >= row || c >= col || r < 0 || c < 0)
        {
            return false;
        }
        
        if (visited[r][c])
        {
            return false;
        }
        
        if (word[idx] != board[r][c])
        {
            return false;
        }
        
        if (idx == word.length() - 1)
        {
            return true;
        }
        
        visited[r][c] = true;
        
        if (run(visited, idx + 1, word, board, r + 1, c)) 
            return true;
        
        if (run(visited, idx + 1, word, board, r - 1, c))
            return true;
        
        if (run(visited, idx + 1, word, board, r, c + 1))
            return true;
        
        if (run(visited, idx + 1, word, board, r, c - 1))
            return true;
        
        visited[r][c] = false;
        
        return false;
    }
    
    bool exist(vector<vector<char>>& board, string word) 
    {
        int row = board.size();
        if (row == 0)
        {
            return false;
        }
        
        int col = board[0].size();
        vector<vector<bool>> visited(row, vector<bool>(col, false));
        
        for (int r = 0; r < row; r++)
        {
            for (int c = 0; c < col; c++)
            {
                if (run(visited, 0, word, board, r, c))
                {
                    return true;
                }
            }
        }
        
        return false;
    }
};
